cmd/ocsp-updater/main.go

```diff
@@ -0,0 +1,202 @@
+	if err != nil {
+		return err
+	}
+
+	timeStamp := time.Now()
+
+	// Record the response.
+	ocspResp := &core.OcspResponse{Serial: serial, CreatedAt: timeStamp, Response: ocspResponse}
+	err = tx.Insert(ocspResp)
+	if err != nil {
+		return err
+	}
+
+	// Reset the update clock
+	status.OCSPLastUpdated = timeStamp
+	_, err = tx.Update(status)
+	if err != nil {
+		return err
+	}
+
+	// Done
+	return nil
+}
+
+func findStaleResponses(cac rpc.CertificateAuthorityClient, dbMap *gorp.DbMap, oldestLastUpdatedTime time.Time, responseLimit int) error {
+	log := blog.GetAuditLogger()
+
+	// If there are fewer than this many days left before the currently-signed
+	// OCSP response expires, sign a new OCSP response.
+	var certificateStatus []core.CertificateStatus
+	_, err := dbMap.Select(&certificateStatus,
+		`SELECT cs.* FROM certificateStatus AS cs
+		 WHERE cs.ocspLastUpdated < ?
+		 ORDER BY cs.ocspLastUpdated ASC
+		 LIMIT ?`, oldestLastUpdatedTime, responseLimit)
+
+	if err == sql.ErrNoRows {
+		log.Info("All up to date. No OCSP responses needed.")
+	} else if err != nil {
+		log.Err(fmt.Sprintf("Error loading certificate status: %s", err))
+	} else {
+		log.Info(fmt.Sprintf("Processing OCSP Responses...\n"))
+		for i, status := range certificateStatus {
+			log.Info(fmt.Sprintf("OCSP %d: %s", i, status.Serial))
+
+			// Each response gets a transaction. To speed this up, we can batch
+			// transactions.
+			tx, err := dbMap.Begin()
+			if err != nil {
+				log.Err(fmt.Sprintf("Error starting transaction, aborting: %s", err))
+				tx.Rollback()
+				return err
+			}
+
+			if err := processResponse(cac, tx, status.Serial); err != nil {
+				log.Err(fmt.Sprintf("Could not process OCSP Response for %s: %s", status.Serial, err))
+				tx.Rollback()
+			} else {
+				log.Info(fmt.Sprintf("OCSP %d: %s OK", i, status.Serial))
+				tx.Commit()
+			}
+		}
+	}
+
+	return err
+}
+
+func main() {
+	app := cmd.NewAppShell("ocsp-updater")
+
+	app.App.Flags = append(app.App.Flags, cli.IntFlag{
+		Name:   "limit",
+		Value:  ocspResponseLimit,
+		EnvVar: "OCSP_LIMIT",
+		Usage:  "Count of responses to process per run",
+	})
+
+	app.Config = func(c *cli.Context, config cmd.Config) cmd.Config {
+		config.OCSP.ResponseLimit = c.GlobalInt("limit")
+		return config
+	}
+
+	app.Action = func(c cmd.Config) {
+		// Set up logging
+		stats, err := statsd.NewClient(c.Statsd.Server, c.Statsd.Prefix)
+		cmd.FailOnError(err, "Couldn't connect to statsd")
+
+		auditlogger, err := blog.Dial(c.Syslog.Network, c.Syslog.Server, c.Syslog.Tag, stats)
+		cmd.FailOnError(err, "Could not connect to Syslog")
+
+		// AUDIT[ Error Conditions ] 9cc4d537-8534-4970-8665-4b382abe82f3
+		defer auditlogger.AuditPanic()
+
+		blog.SetAuditLogger(auditlogger)
+
+		// Configure DB
+		dbMap, err := sa.NewDbMap(c.OCSP.DBDriver, c.OCSP.DBName)
+		cmd.FailOnError(err, "Could not connect to database")
+
+		dbMap.AddTableWithName(core.OcspResponse{}, "ocspResponses").SetKeys(true, "ID")
```

Comment:
Moved the initTables method into database.go and made it always run on a DbMap. This may require some further work when we use a disjoint schema, like that for `certificate-authority-data.go`.